operator/scheduler/experiment_test.go

```diff
@@ -0,0 +1,414 @@
+			existing_resources: []client.Object{},
+			noSchedulerState:   false,
+		},
+		{
+			name: "experiment does not exist in k8s",
+			existing_resources: []client.Object{
+				&mlopsv1alpha1.Experiment{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "foo",
+						Namespace:  "default",
+						Generation: 1,
+					},
+					Spec: mlopsv1alpha1.ExperimentSpec{},
+				},
+			},
+			results: []*scheduler.ExperimentStatusResponse{
+				{
+					ExperimentName:  "foo2",
+					Active:          false,
+					CandidatesReady: false,
+					KubernetesMeta: &scheduler.KubernetesMeta{
+						Namespace:  "default",
+						Generation: 1,
+					},
+				},
+			},
+			noSchedulerState: false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// note that responses_experiments is nill -> scheduler state is not existing
+			var grpcClient mockSchedulerGrpcClient
+			if !test.noSchedulerState {
+				grpcClient = mockSchedulerGrpcClient{
+					responses_subscribe_experiments: test.results,
+					responses_experiments:           test.results,
+				}
+			} else {
+				grpcClient = mockSchedulerGrpcClient{
+					responses_subscribe_experiments: test.results,
+				}
+			}
+			controller := newMockControllerClient(test.existing_resources...)
+			err := controller.SubscribeExperimentEvents(context.Background(), &grpcClient, "")
+			g.Expect(err).To(BeNil())
+
+			isBeingDeleted := map[string]bool{}
+			for _, req := range test.existing_resources {
+				if !req.GetDeletionTimestamp().IsZero() {
+					isBeingDeleted[req.GetName()] = true
+				} else {
+					isBeingDeleted[req.GetName()] = false
+				}
+			}
+			// check that we have reloaded the correct resources if the stata of the scheduler is not correct
+			if test.noSchedulerState {
+				activeResources := 0
+				for idx, req := range test.existing_resources {
+					if req.GetDeletionTimestamp().IsZero() {
+						g.Expect(req.GetName()).To(Equal(grpcClient.requests_experiments[idx].Experiment.GetName()))
+						activeResources++
+					}
+				}
+				g.Expect(len(grpcClient.requests_experiments)).To(Equal(activeResources))
+			} else {
+				g.Expect(len(grpcClient.requests_experiments)).To(Equal(0))
+			}
+
+			// check state is correct for each pipeline
+			for _, r := range test.results {
+				if !isBeingDeleted[r.ExperimentName] {
+					experiment := &mlopsv1alpha1.Experiment{}
+					err := controller.Get(
+						context.Background(),
+						client.ObjectKey{
+							Name:      r.ExperimentName,
+							Namespace: r.GetKubernetesMeta().Namespace,
+						},
+						experiment,
+					)
+
+					// we check if the experiement is not in k8s (existing_resources) then we should not act on it
+					if _, ok := isBeingDeleted[r.ExperimentName]; !ok {
+						g.Expect(err).ToNot(BeNil())
+					} else {
+						g.Expect(err).To(BeNil())
+					}
+
+					if r.CandidatesReady && r.Active && r.MirrorReady {
+						g.Expect(experiment.Status.IsReady()).To(BeTrue())
```

Comment:
I think for those tests is perfectly fine, but in general I propose we move towards `BeTrueBecause` and `BeFalseBecause`, which can print more informative data on test failure. (On failure, `BeTrue` just prints "true" != "false"). [docs](https://onsi.github.io/gomega/#betruebecausereason)

Reply:
fixed in https://github.com/SeldonIO/seldon-core/pull/5726/commits/5946b88c2f1ebf3530a2be56a7682878e16036e4